blackjax/inference/trajectory.py

```diff
@@ -275,6 +313,162 @@ def add_one_state(expansion_state):
     return integrate
 
 
+def dynamic_recursive_integration(
+    integrator: Callable,
+    kinetic_energy: Callable,
+    uturn_check_fn: Callable,
+    divergence_threshold: float,
+    use_robust_uturn_check: bool = False,
+):
+    """Integrate a trajectory and update the proposal recursively in Python
+    until the termination criterion is met.
+
+    This is the implementation of Algorithm 6 from [1] with multinomial sampling.
+
+    Parameters
+    ----------
+    integrator
+        The symplectic integrator used to integrate the hamiltonian trajectory.
+    kinetic_energy
+        Function to compute the current value of the kinetic energy.
+    uturn_check_fn
+        Determines whether the termination criterion has been met.
+    divergence_threshold
+        Value of the difference of energy between two consecutive states above which we say a transition is divergent.
+    use_robust_uturn_check
+        Bool to indicate whether to perform additional U turn check between two trajectory.
+
+    References
+    ----------
+    .. [1]: Hoffman, Matthew D., and Andrew Gelman. "The No-U-Turn sampler: adaptively setting path lengths in Hamiltonian Monte Carlo." J. Mach. Learn. Res. 15.1 (2014): 1593-1623.
+    """
+
+    _, generate_proposal = proposal_generator(kinetic_energy, divergence_threshold)
+    sample_proposal = progressive_uniform_sampling
+
+    def buildtree_integrate(
+        rng_key: jax.numpy.DeviceArray,
+        initial_state: IntegratorState,
+        direction: int,
+        tree_depth: int,
+        step_size,
+        initial_energy: float,
+    ):
```

Comment:
A plain english description of the algorithm, like for `dynamic_multiplicative_expansion` would help lot understanding what's happening below :)

Reply:
added - wdyt?

Reply:
That's enough for someone reading the code. Someone will do another pass one day™ when documenting the module.